### ISSUE TITLE
Generics with bound TypeVars should not default to Any in annotations

### ISSUE THREAD

**Issue by chadrik** [python/mypy#12887]
**Bug Report**

When an unparameterized generic type (is there a better term for this?) is used as an annotation, the generic's bounds (or type choices) are thrown away and provide no protection.  I feel like there should be an implicit bound in this case. 

This is a sharp edge that my team encountered with mypy that allowed a number of issues to go undetected.   The current behavior is not intuitive, and requires workarounds that are conceptually complex. 

**To Reproduce**

```python
from typing import Generic, TypeVar

NumericT = TypeVar('NumericT', int, float)


class BaseThing(Generic[NumericT]):
    def put(self, x: NumericT):
        val = x + 'wrong'    # ERROR:  good, this is expected. 


class IntThing(BaseThing[int]):
    pass


class FloatThing(BaseThing[float]):
    pass


class StringThing(BaseThing[str]):   # ERROR:  good, this is expected. 
    pass


def get() -> BaseThing:
    pass


thing = get()

thing.put('wrong')    # NO ERROR:  this should be an error

if isinstance(thing, IntThing):
    thing.put(2)
```

**Expected Behavior**

mypy should display an error for `thing.put('wrong')` because it is not possible for  a subclass of BaseThing to implement an override of `put()` that accepts a string.

**Actual Behavior**

No error is displayed for `thing.put('wrong')`, though it is displayed for other similar situations.

Output:

```
typevar_bounds.py:8: error: Unsupported operand types for + ("int" and "str")
typevar_bounds.py:8: error: Unsupported operand types for + ("float" and "str")
typevar_bounds.py:19: error: Value of type variable "NumericT" of "BaseThing" cannot be "str"
```

**Workaround**

As far as I can tell, the workaround is to expand my TypeVar to include a Union of the accepted types, then parameterize the return type with that Union:

```python
from typing import Generic, TypeVar, Union

NumericT = TypeVar('NumericT', int, float, Union[int, float])


class BaseThing(Generic[NumericT]):
    def put(self, x: NumericT):
        val = x + 'wrong'


class IntThing(BaseThing[int]):
    pass


class FloatThing(BaseThing[float]):
    pass


def get() -> BaseThing[Union[int, float]]:
    pass


thing = get()
thing.put('wrong')

if isinstance(thing, IntThing):
    thing.put(2)    # ERROR: <nothing> has no attribute "put"
```

Note that the `isinstance` check now gives a new error about `<nothing>`.  In this case I can solve the problem by making `NumericT` covariant. 

I point this out because the solution to this problem unravels into some pretty deep and esoteric typing issues (variance, erased types) and I feel like there's a clear opportunity to do something more intuitve without all this effort on the user's part. 

**Your Environment**

- Mypy version used:  0.960
- Mypy configuration options from `mypy.ini` (and other config files):
  ```
  [mypy]
  warn_no_return = false
  no_implicit_optional = true
  warn_redundant_casts = true
  warn_unused_ignores = true
  ```
- Python version used: 3.8.5
- Operating system and version:  osx

**Comment by erictraut** [comment]
PEP 484 is pretty clear that missing type arguments should default to `Any`. This is true regardless of whether the TypeVar is bound or constrained. So I think mypy is correctly implementing the typing standard.

You can use the `--disallow-any-generics`, and mypy will report missing (implicit `Any`) type arguments for generic types.

**Comment by chadrik** [comment]
I'm happy to bring this issue up in another forum if there's a more appropriate one, but my main question is: does anyone see a fault in my logic?  Do people agree that it's an improvement to use the bounds/constraints to infer a better alternative to Any?

> You can use the --disallow-any-generics, and mypy will report missing (implicit Any) type arguments for generic types.

I turned this on in our code base and it generated another 1k errors, so we won't be turning that on any time soon.  This is overkill anyway.  What I'm looking to improve is the specific situation where allowing it default to Any *throws away protection* that I've already defined in my code, via bounds or constraints.  What I really want is to not have to specify those same protections in numerous places and keep them in sync, e.g. in my TypeVar and also in my type annotations (violates the DRY principle).  If there were a warning I could turn on to detect this case, that'd be a step in the right direction.

**Comment by AlexWaygood** [comment]
> I'm happy to bring this issue up in another forum if there's a more appropriate one, but my main question is: does anyone see a fault in my logic? Do people agree that it's an improvement to use the bounds/constraints to infer a better alternative to Any?

I think your argument is pretty good, but I agree with @erictraut that PEP 484 is pretty clear on this, so this requires a change in the spec. Those discussions need to happen on the [python/typing issue tracker](https://github.com/python/typing/issues) or [the typing-sig mailing list](https://mail.python.org/archives/list/typing-sig@python.org/). Feel free to open a discussion in either of those forums :)

Also, just to note: all the `TypeVar`s in your example are _constrained_ rather than _bound_. https://docs.python.org/3/library/typing.html#typing.TypeVar
